adapters/msft/msft.go

```diff
@@ -0,0 +1,396 @@
+	}
+
+	impExt := impExtOutgoing{
+		Appnexus: impExtOutgoingAppnexus{
+			PlacementID:       ext.Bidder.PlacementId,
+			AllowSmallerSizes: ext.Bidder.AllowSmallerSizes,
+			UsePmtRule:        ext.Bidder.UsePaymentRule,
+			Keywords:          ext.Bidder.Keywords,
+			TrafficSourceCode: ext.Bidder.TrafficSourceCode,
+			PubClick:          ext.Bidder.PubClick,
+			ExtInvCode:        ext.Bidder.ExtInvCode,
+			ExtImpID:          ext.Bidder.ExtImpId,
+		},
+		GPID: ext.GPID,
+	}
+
+	var err error
+	imp.Ext, err = jsonutil.Marshal(impExt)
+
+	return err
+}
+
+func appendMemberId(uri url.URL, memberId int) url.URL {
+	q := uri.Query()
+	q.Set("member_id", fmt.Sprint(memberId))
+	uri.RawQuery = q.Encode()
+	return uri
+}
+
+func (a *adapter) modifyRequestExt(request *openrtb2.BidRequest, requestInfo *adapters.ExtraRequestInfo) error {
+	reqExt, err := getRequestExt(request.Ext)
+	if err != nil {
+		return err
+	}
+
+	reqExtAppnexus, err := a.getAppnexusExt(reqExt, requestInfo.PbsEntryPoint)
+	if err != nil {
+		return err
+	}
+
+	appnexusExtJson, err := jsonutil.Marshal(reqExtAppnexus)
+	if err != nil {
+		return err
+	}
+
+	reqExt["appnexus"] = appnexusExtJson
+
+	request.Ext, err = jsonutil.Marshal(reqExt)
+	return err
+}
+
+func getRequestExt(ext json.RawMessage) (map[string]json.RawMessage, error) {
+	extMap := make(map[string]json.RawMessage)
+
+	if len(ext) > 0 {
+		if err := jsonutil.Unmarshal(ext, &extMap); err != nil {
+			return nil, err
+		}
+	}
+
+	return extMap, nil
+}
+
+func (a *adapter) getAppnexusExt(extMap map[string]json.RawMessage, reqType metrics.RequestType) (requestExAppnexus, error) {
+	var appnexusExt requestExAppnexus
+
+	if appnexusExtJson, exists := extMap["appnexus"]; exists && len(appnexusExtJson) > 0 {
+		if err := jsonutil.Unmarshal(appnexusExtJson, &appnexusExt); err != nil {
+			return appnexusExt, errMalformedRequestExt
+		}
+	}
+
+	if prebidJson, exists := extMap["prebid"]; exists {
+		_, valueType, _, err := jsonparser.Get(prebidJson, "targeting", "includebrandcategory")
+		if err != nil && !errors.Is(err, jsonparser.KeyPathNotFoundError) {
+			return appnexusExt, err
+		}
+
+		if valueType == jsonparser.Object {
+			appnexusExt.BrandCategoryUniqueness = ptrutil.ToPtr(true)
+			appnexusExt.IncludeBrandCategory = ptrutil.ToPtr(true)
+		}
+	}
+
+	if reqType == metrics.ReqTypeAMP {
+		appnexusExt.IsAMP = 1
+	}
+
+	if reqType == metrics.ReqTypeVideo {
+		appnexusExt.HeaderBiddingSource = a.hbSourceVideo
+	} else {
+		appnexusExt.HeaderBiddingSource = a.hbSource
+	}
+
+	return appnexusExt, nil
+}
+
+func buildRequests(imps []openrtb2.Imp, request *openrtb2.BidRequest, uri string) ([]*adapters.RequestData, []error) {
+	var (
+		requestsCount = (len(imps) + maxImpsPerReq - 1) / maxImpsPerReq
+		requests      = make([]*adapters.RequestData, 0, requestsCount)
+	)
+
+	for i := 0; i < requestsCount; i++ {
+		imps := imps[i*maxImpsPerReq : min((i+1)*maxImpsPerReq, len(imps))]
```

Comment:
You're shadowing the function parameter. This works. Do you want to rename it to something else though?

Reply:
Sure